### ISSUE TITLE
Local two-stage evaluation

### ISSUE THREAD

**Issue by hidanom** [autonomousvision/navsim#81]
Hello,

I have two questions regarding the two-stage evaluation process.

1. The traffic policies are supposed to be different for the two stages, but the eval script (`run_pdm_score.py`) seems to apply the same given policy to all of the scenes, both real and synthetic. How can I configure the evaluation correctly?
2. More generally, exactly how am I supposed to reproduce the leaderboard warmup evaluation locally? I've tried running the CV agent evaluation on the warmup test set, but some of the resulting metrics diverge from what's in the leaderboard (e.g., `"ego_progress"`). Could you provide the specific evaluation command with all of the required config overrides?

Thanks!

**Comment by VVeiCao** [comment]
Hi @hidanom,

For the warmup leaderboard, we currently do not support different traffic agent settings. The synthetic sensor data provided is also based on non-reactive traffic agents.

For the CV agent, if you run it using the following setup, you should obtain the same results as those on the leaderboard:
```
TRAIN_TEST_SPLIT=warmup_navsafe_two_stage_extended
python $NAVSIM_DEVKIT_ROOT/navsim/planning/script/run_pdm_score.py \
    train_test_split=$TRAIN_TEST_SPLIT \
    experiment_name= YOUR_EXP \
    traffic_agents_policy=log_replay_traffic_agents \
    agent=constant_velocity_agent \
    metric_cache_path= YOUR_CACHE \
```
Best,
Wei

**Comment by hidanom** [comment]
I've tried running this and the resulting `"ego_progress"` metric diverges from the leaderboard value.
[cv_original.csv](https://github.com/user-attachments/files/19473612/cv_original.csv)
Do you get a different result running the evaluation locally?

Also, do you plan to use two separate traffic policies for the private two-stage evaluation? If so, when are you going to support this sort of evaluation in the repo?

**Comment by VVeiCao** [comment]
Hi @hidanom，

Yes, on my side the results are the same as those on the leaderboard. Could you try recloning the code and clearing the metric cache to see if that helps?

[2025.03.26.18.39.26.csv](https://github.com/user-attachments/files/19473953/2025.03.26.18.39.26.csv)

Regarding your second question — yes, different traffic policies will be available in Navsim 2.1. We haven’t finalized the release date yet.

Best,
Wei

**Comment by hidanom** [comment]
Interesting. The code seems to be identical to the main of this repo and I do re-cache everything. Which version of `shapely` are you running locally?

By the way, `requirements.txt` doesn't seem to be strict enough. Running an older scipy version (was it 1.9?) messes up calculation of metric cache. Somewhere around here
https://github.com/autonomousvision/navsim/blob/4de93c5679d6566e4225d83cf2132eb136824cbc/navsim/planning/metric_caching/metric_cache_processor.py#L165
some of the collected tracks only contain a single state and scipy throws an error trying to instantiate an interpolator. Newer version (1.13) is more lenient so the interpolation pipeline seems to work. Although, I'm not sure it processes these single-state tracks reasonably:
https://github.com/autonomousvision/navsim/blob/4de93c5679d6566e4225d83cf2132eb136824cbc/navsim/planning/metric_caching/metric_cache_processor.py#L177
seems to be copying the single available state to all of the new timestamps, while a longer track would only get interpolated tracks between `start_time` and `end_time`.

**Comment by VVeiCao** [comment]
Hi @hidanom,

Thanks for pointing out. I think you referred to `scipy` instead of `shapely`, right?
Locally we have version 1.13.1 for `scipy`.

**Comment by hidanom** [comment]
No, I meant `shapely`. Ego progress metric relies on it to project proposed trajectory onto the centerline. Using different versions might explain the divergence in the metric.

**Comment by VVeiCao** [comment]
`Shapely` should be version >=2.0.0 — it's neither 1.9 nor 1.13 (written in `requirements.txt`). 
Could you provide the exact versions of `Shapely` where you noticed the projection differences?

**Comment by VVeiCao** [comment]
Hi @hidanom,

If there is no further question, I will close this off. Feel free to reopen.

Best,
Wei
